File: algorithmic-strategies/Greedy Algorithms/C++/code.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
#include <limits>
#include <string>
#include <functional>

using namespace std;

struct Activity
{
    int start;
    int finish;
};

int activitySelection(vector<Activity> activities)
{
    if (activities.empty())
    {
        return 0;
    }

    sort(
        activities.begin(),
        activities.end(),
        [](const Activity& a, const Activity& b)
        {
            return a.finish < b.finish;
        }
    );

    int selected = 1;
    int lastFinish = activities[0].finish;

    for (size_t i = 1; i < activities.size(); i++)
    {
        if (activities[i].start >= lastFinish)
        {
            selected++;
            lastFinish = activities[i].finish;
        }
    }

    return selected;
}

struct Item
{
    double value;
    double weight;
};

double fractionalKnapsack(
    vector<Item> items,
    double capacity
)
{
    sort(
        items.begin(),
        items.end(),
        [](const Item& a, const Item& b)
        {
            return (a.value / a.weight) >
                   (b.value / b.weight);
        }
    );

    double totalValue = 0.0;

    for (const Item& item : items)
    {
        if (capacity <= 0)
        {
            break;
        }

        if (item.weight <= capacity)
        {
            totalValue += item.value;
            capacity -= item.weight;
        }
        else
        {
            double fraction =
                capacity / item.weight;

            totalValue += item.value * fraction;
            capacity = 0;
        }
    }

    return totalValue;
}

struct Job
{
    char id;
    int deadline;
    int profit;
};

pair<vector<Job>, int> jobSequencing(
    vector<Job> jobs
)
{
    sort(
        jobs.begin(),
        jobs.end(),
        [](const Job& a, const Job& b)
        {
            return a.profit > b.profit;
        }
    );

    int maxDeadline = 0;

    for (const Job& job : jobs)
    {
        maxDeadline =
            max(maxDeadline, job.deadline);
    }

    vector<Job> slots(maxDeadline);
    vector<bool> occupied(maxDeadline, false);

    vector<Job> selected;
    int totalProfit = 0;

    for (const Job& job : jobs)
    {
        for (int slot = job.deadline - 1;
             slot >= 0;
             slot--)
        {
            if (!occupied[slot])
            {
                occupied[slot] = true;
                slots[slot] = job;

                selected.push_back(job);
                totalProfit += job.profit;

                break;
            }
        }
    }

    return {selected, totalProfit};
}

struct HuffmanNode
{
    char character;
    int frequency;

    HuffmanNode* left;
    HuffmanNode* right;

    HuffmanNode(
        char character,
        int frequency
    )
        : character(character),
          frequency(frequency),
          left(nullptr),
          right(nullptr)
    {
    }
};

struct CompareHuffman
{
    bool operator()(
        const HuffmanNode* a,
        const HuffmanNode* b
    ) const
    {
        return a->frequency > b->frequency;
    }
};

void generateHuffmanCodes(
    HuffmanNode* root,
    const string& code,
    vector<pair<char, string>>& codes
)
{
    if (root == nullptr)
    {
        return;
    }

    if (root->left == nullptr &&
        root->right == nullptr)
    {
        codes.push_back({
            root->character,
            code.empty() ? "0" : code
        });

        return;
    }

    generateHuffmanCodes(
        root->left,
        code + "0",
        codes
    );

    generateHuffmanCodes(
        root->right,
        code + "1",
        codes
    );
}

vector<pair<char, string>> huffmanCodes(
    const vector<char>& characters,
    const vector<int>& frequencies
)
{
    priority_queue<
        HuffmanNode*,
        vector<HuffmanNode*>,
        CompareHuffman
    > minHeap;

    for (size_t i = 0;
         i < characters.size();
         i++)
    {
        minHeap.push(
            new HuffmanNode(
                characters[i],
                frequencies[i]
            )
        );
    }

    while (minHeap.size() > 1)
    {
        HuffmanNode* left =
            minHeap.top();

        minHeap.pop();

        HuffmanNode* right =
            minHeap.top();

        minHeap.pop();

        HuffmanNode* parent =
            new HuffmanNode(
                '\0',
                left->frequency +
                right->frequency
            );

        parent->left = left;
        parent->right = right;

        minHeap.push(parent);
    }

    HuffmanNode* root = minHeap.top();

    vector<pair<char, string>> codes;

    generateHuffmanCodes(
        root,
        "",
        codes
    );

    function<void(HuffmanNode*)> cleanup =
        [&](HuffmanNode* node)
        {
            if (node == nullptr)
            {
                return;
            }

            cleanup(node->left);
            cleanup(node->right);

            delete node;
        };

    cleanup(root);

    return codes;
}

struct Edge
{
    int source;
    int destination;
    int weight;
};

struct Graph
{
    int vertices;
    vector<vector<pair<int, int>>> adjacency;
};

class DisjointSet
{
private:
    vector<int> parent;
    vector<int> rank;

public:
    explicit DisjointSet(int size)
        : parent(size),
          rank(size, 0)
    {
        for (int i = 0; i < size; i++)
        {
            parent[i] = i;
        }
    }

    int find(int vertex)
    {
        if (parent[vertex] != vertex)
        {
            parent[vertex] =
                find(parent[vertex]);
        }

        return parent[vertex];
    }

    void unite(int a, int b)
    {
        int rootA = find(a);
        int rootB = find(b);

        if (rootA == rootB)
        {
            return;
        }

        if (rank[rootA] < rank[rootB])
        {
            parent[rootA] = rootB;
        }
        else if (rank[rootA] > rank[rootB])
        {
            parent[rootB] = rootA;
        }
        else
        {
            parent[rootB] = rootA;
            rank[rootA]++;
        }
    }
};

vector<Edge> kruskalMST(
    int vertexCount,
    vector<Edge> edges
)
{
    sort(
        edges.begin(),
        edges.end(),
        [](const Edge& a, const Edge& b)
        {
            return a.weight < b.weight;
        }
    );

    DisjointSet sets(vertexCount);

    vector<Edge> mst;

    for (const Edge& edge : edges)
    {
        if (sets.find(edge.source) !=
            sets.find(edge.destination))
        {
            mst.push_back(edge);

            sets.unite(
                edge.source,
                edge.destination
            );

            if (mst.size() ==
                static_cast<size_t>(vertexCount - 1))
            {
                break;
            }
        }
    }

    return mst;
}

vector<Edge> primMST(
    const Graph& graph,
    int start
)
{
    const int INF =
        numeric_limits<int>::max();

    vector<int> key(
        graph.vertices,
        INF
    );

    vector<int> parent(
        graph.vertices,
        -1
    );

    vector<bool> included(
        graph.vertices,
        false
    );

    key[start] = 0;

    vector<Edge> mst;

    for (int count = 0;
         count < graph.vertices;
         count++)
    {
        int vertex = -1;
        int minimum = INF;

        for (int v = 0;
             v < graph.vertices;
             v++)
        {
            if (!included[v] &&
                key[v] < minimum)
            {
                minimum = key[v];
                vertex = v;
            }
        }

        if (vertex == -1)
        {
            break;
        }

        included[vertex] = true;

        if (parent[vertex] != -1)
        {
            mst.push_back({
                parent[vertex],
                vertex,
                key[vertex]
            });
        }

        for (const auto& edge :
             graph.adjacency[vertex])
        {
            int neighbour = edge.first;
            int weight = edge.second;

            if (!included[neighbour] &&
                weight < key[neighbour])
            {
                key[neighbour] = weight;
                parent[neighbour] = vertex;
            }
        }
    }

    return mst;
}

vector<int> dijkstra(
    const Graph& graph,
    int source
)
{
    const int INF =
        numeric_limits<int>::max();

    vector<int> distances(
        graph.vertices,
        INF
    );

    priority_queue<
        pair<int, int>,
        vector<pair<int, int>>,
        greater<pair<int, int>>
    > minHeap;

    distances[source] = 0;

    minHeap.push({
        0,
        source
    });

    while (!minHeap.empty())
    {
        int distance =
            minHeap.top().first;

        int vertex =
            minHeap.top().second;

        minHeap.pop();

        if (distance != distances[vertex])
        {
            continue;
        }

        for (const auto& edge :
             graph.adjacency[vertex])
        {
            int neighbour = edge.first;
            int weight = edge.second;

            if (weight < 0)
            {
                continue;
            }

            int newDistance =
                distance + weight;

            if (newDistance <
                distances[neighbour])
            {
                distances[neighbour] =
                    newDistance;

                minHeap.push({
                    newDistance,
                    neighbour
                });
            }
        }
    }

    return distances;
}

void printMST(
    const vector<Edge>& mst
)
{
    int totalWeight = 0;

    for (const Edge& edge : mst)
    {
        cout
            << edge.source
            << " -- "
            << edge.destination
            << " ("
            << edge.weight
            << ")\n";

        totalWeight += edge.weight;
    }

    cout
        << "Total weight: "
        << totalWeight
        << '\n';
}

int main()
{

    vector<Activity> activities = {
        {1, 3},
        {2, 4},
        {3, 5},
        {5, 7},
        {8, 9},
        {5, 9}
    };

    cout
        << "Activity Selection: "
        << activitySelection(activities)
        << " activities\n";

    vector<Item> items = {
        {60, 10},
        {100, 20},
        {120, 30}
    };

    cout
        << "Fractional Knapsack: "
        << fractionalKnapsack(items, 50)
        << '\n';

    vector<Job> jobs = {
        {'A', 2, 100},
        {'B', 1, 19},
        {'C', 2, 27},
        {'D', 1, 25},
        {'E', 3, 15}
    };

    auto jobResult =
        jobSequencing(jobs);

    cout << "Job Sequencing: ";

    for (const Job& job :
         jobResult.first)
    {
        cout << job.id << ' ';
    }

    cout
        << "\nTotal profit: "
        << jobResult.second
        << '\n';

    vector<char> characters = {
        'A', 'B', 'C',
        'D', 'E', 'F'
    };

    vector<int> frequencies = {
        5, 9, 12,
        13, 16, 45
    };

    auto codes =
        huffmanCodes(
            characters,
            frequencies
        );

    cout << "Huffman Codes:\n";

    for (const auto& code : codes)
    {
        cout
            << code.first
            << ": "
            << code.second
            << '\n';
    }

    const int vertexCount = 6;

    vector<Edge> edges = {
        {0, 1, 4},
        {0, 2, 2},
        {1, 2, 1},
        {1, 3, 5},
        {2, 3, 8},
        {2, 4, 10},
        {3, 4, 2},
        {3, 5, 6},
        {4, 5, 3}
    };

    vector<Edge> kruskalResult =
        kruskalMST(
            vertexCount,
            edges
        );

    cout << "\nKruskal MST:\n";

    printMST(kruskalResult);

    Graph graph;

    graph.vertices = vertexCount;

    graph.adjacency.resize(
        vertexCount
    );

    auto addUndirectedEdge =
        [&](int source,
            int destination,
            int weight)
        {
            graph.adjacency[source].push_back({
                destination,
                weight
            });

            graph.adjacency[destination].push_back({
                source,
                weight
            });
        };

    addUndirectedEdge(0, 1, 4);
    addUndirectedEdge(0, 2, 2);
    addUndirectedEdge(1, 2, 1);
    addUndirectedEdge(1, 3, 5);
    addUndirectedEdge(2, 3, 8);
    addUndirectedEdge(2, 4, 10);
    addUndirectedEdge(3, 4, 2);
    addUndirectedEdge(3, 5, 6);
    addUndirectedEdge(4, 5, 3);

    vector<Edge> primResult =
        primMST(
            graph,
            0
        );

    cout << "\nPrim MST:\n";

    printMST(primResult);

    vector<int> distances =
        dijkstra(
            graph,
            0
        );

    cout
        << "\nDijkstra from vertex 0:\n";

    for (int i = 0;
         i < vertexCount;
         i++)
    {
        cout
            << i
            << ": ";

        if (distances[i] ==
            numeric_limits<int>::max())
        {
            cout << "unreachable";
        }
        else
        {
            cout << distances[i];
        }

        cout << '\n';
    }

    return 0;
}

File: algorithmic-strategies/Greedy Algorithms/Python/code.py
from dataclasses import dataclass
import heapq

@dataclass
class Activity:
    start: int
    finish: int


def activity_selection(activities):
    if not activities:
        return 0

    activities.sort(key=lambda activity: activity.finish)

    selected = 1
    last_finish = activities[0].finish

    for activity in activities[1:]:
        if activity.start >= last_finish:
            selected += 1
            last_finish = activity.finish

    return selected

@dataclass
class Item:
    value: float
    weight: float


def fractional_knapsack(items, capacity):
    items.sort(
        key=lambda item: item.value / item.weight,
        reverse=True
    )

    total_value = 0.0

    for item in items:
        if capacity <= 0:
            break

        if item.weight <= capacity:
            total_value += item.value
            capacity -= item.weight
        else:
            fraction = capacity / item.weight
            total_value += item.value * fraction
            capacity = 0

    return total_value

@dataclass
class Job:
    id: str
    deadline: int
    profit: int


def job_sequencing(jobs):
    jobs.sort(
        key=lambda job: job.profit,
        reverse=True
    )

    max_deadline = max(
        (job.deadline for job in jobs),
        default=0
    )

    slots = [None] * max_deadline
    total_profit = 0
    selected = []

    for job in jobs:
        for slot in range(job.deadline - 1, -1, -1):
            if slots[slot] is None:
                slots[slot] = job
                selected.append(job)
                total_profit += job.profit
                break

    return selected, total_profit

@dataclass
class HuffmanNode:
    frequency: int
    character: str = ""
    left: object = None
    right: object = None


def build_huffman_tree(characters, frequencies):
    heap = []
    counter = 0

    for character, frequency in zip(
        characters,
        frequencies
    ):
        node = HuffmanNode(
            frequency=frequency,
            character=character
        )

        heapq.heappush(
            heap,
            (frequency, counter, node)
        )

        counter += 1

    if not heap:
        return None

    while len(heap) > 1:
        _, _, left = heapq.heappop(heap)
        _, _, right = heapq.heappop(heap)

        parent = HuffmanNode(
            frequency=left.frequency + right.frequency,
            left=left,
            right=right
        )

        heapq.heappush(
            heap,
            (
                parent.frequency,
                counter,
                parent
            )
        )

        counter += 1

    return heapq.heappop(heap)[2]


def generate_huffman_codes(
    root,
    code="",
    codes=None
):
    if codes is None:
        codes = {}

    if root is None:
        return codes

    if root.left is None and root.right is None:
        codes[root.character] = (
            code if code else "0"
        )

        return codes

    generate_huffman_codes(
        root.left,
        code + "0",
        codes
    )

    generate_huffman_codes(
        root.right,
        code + "1",
        codes
    )

    return codes


def huffman_codes(characters, frequencies):
    root = build_huffman_tree(
        characters,
        frequencies
    )

    return generate_huffman_codes(root)

@dataclass
class Edge:
    source: int
    destination: int
    weight: int


class Graph:
    def __init__(self, vertices):
        self.vertices = vertices
        self.adjacency = [
            [] for _ in range(vertices)
        ]

    def add_edge(
        self,
        source,
        destination,
        weight
    ):
        self.adjacency[source].append(
            (destination, weight)
        )

        self.adjacency[destination].append(
            (source, weight)
        )

class DisjointSet:
    def __init__(self, size):
        self.parent = list(range(size))
        self.rank = [0] * size

    def find(self, vertex):
        if self.parent[vertex] != vertex:
            self.parent[vertex] = self.find(
                self.parent[vertex]
            )

        return self.parent[vertex]

    def union(self, a, b):
        root_a = self.find(a)
        root_b = self.find(b)

        if root_a == root_b:
            return

        if self.rank[root_a] < self.rank[root_b]:
            self.parent[root_a] = root_b

        elif self.rank[root_a] > self.rank[root_b]:
            self.parent[root_b] = root_a

        else:
            self.parent[root_b] = root_a
            self.rank[root_a] += 1

def kruskal_mst(vertices, edges):
    edges.sort(
        key=lambda edge: edge.weight
    )

    disjoint_set = DisjointSet(vertices)

    mst = []

    for edge in edges:
        if (
            disjoint_set.find(edge.source)
            != disjoint_set.find(edge.destination)
        ):
            mst.append(edge)

            disjoint_set.union(
                edge.source,
                edge.destination
            )

            if len(mst) == vertices - 1:
                break

    return mst

def prim_mst(graph, start):
    visited = [False] * graph.vertices

    min_heap = [
        (0, start, -1)
    ]

    mst = []

    while min_heap:
        weight, vertex, parent = heapq.heappop(
            min_heap
        )

        if visited[vertex]:
            continue

        visited[vertex] = True

        if parent != -1:
            mst.append(
                Edge(
                    parent,
                    vertex,
                    weight
                )
            )

        for neighbour, edge_weight in graph.adjacency[vertex]:
            if not visited[neighbour]:
                heapq.heappush(
                    min_heap,
                    (
                        edge_weight,
                        neighbour,
                        vertex
                    )
                )

    return mst

def dijkstra(graph, source):
    infinity = float("inf")

    distances = [
        infinity
        for _ in range(graph.vertices)
    ]

    distances[source] = 0

    min_heap = [
        (0, source)
    ]

    while min_heap:
        distance, vertex = heapq.heappop(
            min_heap
        )

        if distance != distances[vertex]:
            continue

        for neighbour, weight in graph.adjacency[vertex]:
            if weight < 0:
                raise ValueError(
                    "Dijkstra's algorithm "
                    "does not support negative weights."
                )

            new_distance = distance + weight

            if new_distance < distances[neighbour]:
                distances[neighbour] = new_distance

                heapq.heappush(
                    min_heap,
                    (
                        new_distance,
                        neighbour
                    )
                )

    return distances

def print_mst(mst):
    total_weight = 0

    for edge in mst:
        print(
            f"{edge.source} -- "
            f"{edge.destination} "
            f"({edge.weight})"
        )

        total_weight += edge.weight

    print(
        f"Total weight: {total_weight}"
    )

if __name__ == "__main__":

    activities = [
        Activity(1, 3),
        Activity(2, 4),
        Activity(3, 5),
        Activity(5, 7),
        Activity(8, 9),
        Activity(5, 9)
    ]

    print(
        "Activity Selection:",
        activity_selection(activities),
        "activities"
    )

    items = [
        Item(60, 10),
        Item(100, 20),
        Item(120, 30)
    ]

    print(
        "Fractional Knapsack:",
        fractional_knapsack(items, 50)
    )

    jobs = [
        Job("A", 2, 100),
        Job("B", 1, 19),
        Job("C", 2, 27),
        Job("D", 1, 25),
        Job("E", 3, 15)
    ]

    selected, total_profit = job_sequencing(
        jobs
    )

    print(
        "Job Sequencing:",
        [job.id for job in selected]
    )

    print(
        "Total profit:",
        total_profit
    )

    characters = [
        "A", "B", "C",
        "D", "E", "F"
    ]

    frequencies = [
        5, 9, 12,
        13, 16, 45
    ]

    codes = huffman_codes(
        characters,
        frequencies
    )

    print("Huffman Codes:")

    for character, code in codes.items():
        print(
            f"{character}: {code}"
        )

    vertices = 6

    edges = [
        Edge(0, 1, 4),
        Edge(0, 2, 2),
        Edge(1, 2, 1),
        Edge(1, 3, 5),
        Edge(2, 3, 8),
        Edge(2, 4, 10),
        Edge(3, 4, 2),
        Edge(3, 5, 6),
        Edge(4, 5, 3)
    ]

    kruskal_result = kruskal_mst(
        vertices,
        edges.copy()
    )

    print("\nKruskal MST:")

    print_mst(kruskal_result)

    graph = Graph(vertices)

    graph.add_edge(0, 1, 4)
    graph.add_edge(0, 2, 2)
    graph.add_edge(1, 2, 1)
    graph.add_edge(1, 3, 5)
    graph.add_edge(2, 3, 8)
    graph.add_edge(2, 4, 10)
    graph.add_edge(3, 4, 2)
    graph.add_edge(3, 5, 6)
    graph.add_edge(4, 5, 3)

    prim_result = prim_mst(
        graph,
        0
    )

    print("\nPrim MST:")

    print_mst(prim_result)

    distances = dijkstra(
        graph,
        0
    )

    print(
        "\nDijkstra from vertex 0:"
    )

    for vertex, distance in enumerate(
        distances
    ):
        if distance == float("inf"):
            print(
                f"{vertex}: unreachable"
            )
        else:
            print(
                f"{vertex}: {distance}"
            )